### ISSUE TITLE
TF 0.15 module with multiple openstack providers

### ISSUE THREAD

**Issue by per-lind** [hashicorp/terraform#28407]
Hi
Not sure if this is related to the openstack provider or terraform itself, but it feels more terraform related.
Trying to update a module per the warning we got when upgrading to tf 0.15 reults in Error.

### Terraform Version

```
Terraform v0.15.0
on linux_amd64
+ provider registry.terraform.io/terraform-provider-openstack/openstack v1.40.0
```

### Terraform Configuration Files

https://github.com/per-lind/tf-demo

### Debug Output

https://gist.github.com/per-lind/adfedc5a14441833738f69dfcec9a0e1
### Expected Behavior

I expected the new registry.terraform.io/terraform-provider-openstack/openstack provider to be used and no reference to the old registry.terraform.io/hashicorp/openstack
### Actual Behavior

```
 Error: Invalid type for provider module.multi_rule.provider["registry.terraform.io/terraform-provider-openstack/openstack"].Sto2
│ 
│   on main.tf line 34, in module "multi_rule":
│   34:     openstack.Sto2   = openstack.Sto2
│ 
│ Cannot use configuration from
│ provider["registry.terraform.io/hashicorp/openstack"].Sto2 for
│ module.multi_rule.provider["registry.terraform.io/terraform-provider-openstack/openstack"].Sto2.
│ The given provider configuration is for a different provider type.
```

### Steps to Reproduce

1. `terraform init`

To get the initial warning remove commented providers in modules/test/main.tf and comment line 5 in the same file. https://github.com/per-lind/tf-demo/tree/before-upgade

**Comment by jbardin** [comment]
Thanks for filing the issue @per-lind! I can indeed confirm there is something strange being triggered by your example.

The provider configurations in the root module are being detected as `provider["registry.terraform.io/hashicorp/openstack"]` rather than `provider["registry.terraform.io/terraform-provider-openstack/openstack"]`, even with the appropriate `required_providers` set.

**Comment by dimisjim** [comment]
hey @jbardin 

Is this released with 0.15.1?
